### PR TITLE
feat: allow single parameter dns zone policy

### PR DESCRIPTION
**Pull request by jaredfholgate**
## Overview/Summary

A backwards compatible update to the Deploy-Private-DNS-Zones policy set definition.

This enables passing a subscription id, resource group name and region to automatically generate the resource ids for the policy definitions.

## This PR fixes/adds/changes/removes

N/A

### Breaking Changes

None, designed to be backwards compatible

## Testing Evidence

Please provide any testing evidence to show that your Pull Request works/fixes as described and planned (include screenshots, if appropriate).

### Testing URLs

The below URLs can be updated where the placeholders are, look for `{YOUR GITHUB BRANCH NAME HERE - Remove Curly Brackets Also}` & `{YOUR GITHUB BRANCH NAME HERE - Remove Curly Brackets Also}`, to allow you to test your portal deployment experience.

> Please also replace the curly brackets on the placeholders `{}`

#### Azure Public

[Deploy To Azure](https://portal.azure.com/#blade/Microsoft_Azure_CreateUIDef/CustomDeploymentBlade/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2FEnterprise-Scale%2Ffeat-allow-single-parameter-dns-zone-policy%2FeslzArm%2FeslzArm.json/uiFormDefinitionUri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2FEnterprise-Scale%2Ffeat-allow-single-parameter-dns-zone-policy%2FeslzArm%2Feslz-portal.json)

#### Azure US Gov (Fairfax)
[Deploy To Azure](https://portal.azure.us/#blade/Microsoft_Azure_CreateUIDef/CustomDeploymentBlade/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2FEnterprise-Scale%2Ffeat-allow-single-parameter-dns-zone-policy%2FeslzArm%2FeslzArm.json/uiFormDefinitionUri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2FEnterprise-Scale%2Ffeat-allow-single-parameter-dns-zone-policy%2FeslzArm%2Ffairfaxeslz-portal.json)

## As part of this Pull Request I have

- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/Enterprise-Scale/pulls)
- [x] Associated it with relevant [issues](https://github.com/Azure/Enterprise-Scale/issues), for tracking and closure.
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/Enterprise-Scale/tree/main)
- [x] Performed testing and provided evidence.
- [x] Ensured [contribution guidance](https://github.com/Azure/Enterprise-Scale/wiki/ALZ-Contribution-Guide) is followed.
- [x] Updated relevant and associated documentation.
- [x] Updated the ["What's New?"](https://github.com/Azure/Enterprise-Scale/wiki/Whats-new) wiki page (located: `/docs/wiki/whats-new.md`)

### CODE REVIEW

File: src/resources/Microsoft.Authorization/policySetDefinitions/Deploy-Private-DNS-Zones.json [hunk 1]
```diff
@@ -8,14 +8,192 @@
...
+        "defaultValue": "",
+        "metadata": {
+          "displayName": "Region",
+          "description": "The region where the private DNS zones are deployed. If this is specified, it will override any individual private DNS zone resource ids specified."
+        }
+      },
+      "dnzZoneRegionShortNames": {
+        "type": "object",
+        "defaultValue": {
+          "uaenorth" : "uan",
```
Comment (matt-FFFFFF) [inline src/resources/Microsoft.Authorization/policySetDefinitions/Deploy-Private-DNS-Zones.json:54]: nit - is it worth sorting these for ease of maintenance?

File: src/resources/Microsoft.Authorization/policySetDefinitions/Deploy-Private-DNS-Zones.json [hunk 2]
```diff
@@ -8,14 +8,192 @@
...
+        },
+        "metadata": {
+          "displayName": "Region Short Name Mapping",
+          "description": "Mapping of region to private DNS zone resource id. If the region is not specified, the default private DNS zone resource id will be used."
+        }
+      },
+      "dnsZoneNames": {
+        "type": "object",
+        "defaultValue": {
+          "azureFilePrivateDnsZoneId": "privatelink.afs.azure.net",
```
Comment (matt-FFFFFF) [inline src/resources/Microsoft.Authorization/policySetDefinitions/Deploy-Private-DNS-Zones.json:120]: again is it worth sorting these?

File: src/resources/Microsoft.Authorization/policySetDefinitions/Deploy-Private-DNS-Zones.json [hunk 3]
```diff
@@ -592,29 +770,29 @@
           "description": "Private DNS Zone Identifier"
         }
       },
-      "azureSiteRecoveryBackupPrivateDnsZoneID": {
+      "azureSiteRecoveryBackupPrivateDnsZoneId": {
         "type": "string",
         "defaultValue": "",
         "metadata": {
-          "displayName": "azureSiteRecoveryBackupPrivateDnsZoneID",
+          "displayName": "azureSiteRecoveryBackupPrivateDnsZoneId",
```
Comment (matt-FFFFFF) [inline src/resources/Microsoft.Authorization/policySetDefinitions/Deploy-Private-DNS-Zones.json:777]: 👏 👏 👏 

Love this

**Comment by matt-FFFFFF** [comment]
I get this error when trying to deploy the policy set. Investigating...


`Bad Request({"error":{"code":"InvalidPolicyParameters","message":"A function or parameter in policy set or assignment 'Deploy-Private-DNS-Zones' associated with the policy definition '06695360-db88-47f6-b976-7500d4297475' could not be validated. If using template functions, try following the tips in: https://aka.ms/policy-avoiding-template-failures. The inner exception 'The language expression property '' doesn't exist, available properties are 'uaenorth, northcentralus, malaysiawest, eastus, uksouth, westcentralus, israelcentral, southeastasia, malaysiasouth, koreacentral, northeurope, australiaeast, southafricanorth, norwaywest, norwayeast, westus3, eastus2euap, centralus, mexicocentral, canadacentral, japaneast, swedencentral, taiwannorth, germanynorth, centralindia, westindia, newzealandnorth, australiacentral, ukwest, germanywestcentral, brazilsouth, francecentral, brazilsoutheast, westus2, eastus2, centraluseuap, australiacentral2, francesouth, southafricawest, koreasouth, southindia, canadaeast, qatarcentral, spaincentral, westeurope, japanwest, southcentralus, polandcentral, switzerlandwest, australiasoutheast, switzerlandnorth, italynorth, uaecentral, eastasia, chilecentral, westus, swedensouth'.'."}})`

**Comment by jaredfholgate** [comment]
> I get this error when trying to deploy the policy set. Investigating...
> 
> `Bad Request({"error":{"code":"InvalidPolicyParameters","message":"A function or parameter in policy set or assignment 'Deploy-Private-DNS-Zones' associated with the policy definition '06695360-db88-47f6-b976-7500d4297475' could not be validated. If using template functions, try following the tips in: https://aka.ms/policy-avoiding-template-failures. The inner exception 'The language expression property '' doesn't exist, available properties are 'uaenorth, northcentralus, malaysiawest, eastus, uksouth, westcentralus, israelcentral, southeastasia, malaysiasouth, koreacentral, northeurope, australiaeast, southafricanorth, norwaywest, norwayeast, westus3, eastus2euap, centralus, mexicocentral, canadacentral, japaneast, swedencentral, taiwannorth, germanynorth, centralindia, westindia, newzealandnorth, australiacentral, ukwest, germanywestcentral, brazilsouth, francecentral, brazilsoutheast, westus2, eastus2, centraluseuap, australiacentral2, francesouth, southafricawest, koreasouth, southindia, canadaeast, qatarcentral, spaincentral, westeurope, japanwest, southcentralus, polandcentral, switzerlandwest, australiasoutheast, switzerlandnorth, italynorth, uaecentral, eastasia, chilecentral, westus, swedensouth'.'."}})`

Thanks for testing it. I will take a look and try to fix it. :)

**Review by matt-FFFFFF — DISMISSED** [review]

**Review by Springstone — DISMISSED** [review]
Busy testing deployment. Please add a meaningful update to What's New wiki.

File: utils/policyTesting/Install-PolicySetDefinitions.ps1
Comment (jtracey93) [inline utils/policyTesting/Install-PolicySetDefinitions.ps1:1]: Whats this for? I can see it deploy the policy set defs to an MG, but just wondering where its used?

**Review by jtracey93 — CHANGES_REQUESTED** [review]

File: utils/policyTesting/Install-PolicySetDefinitions.ps1
Comment (jaredfholgate) [reply utils/policyTesting/Install-PolicySetDefinitions.ps1:1]: As mentioned, it is for local testing purposes. Can add equivalent for policy definitions and policy assignments too eventually.

File: utils/policyTesting/Install-PolicySetDefinitions.ps1 [hunk 4]
```diff
@@ -0,0 +1,32 @@
+param(
+  $managementGroupId = "6da17bdb-c2cc-4a35-8f10-cded38ebfc47",
```
Comment (jtracey93) [inline utils/policyTesting/Install-PolicySetDefinitions.ps1:2]:
```suggestion
  $managementGroupId = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx",
```

**Review by jtracey93 — CHANGES_REQUESTED** [review]

**Review by jtracey93 — APPROVED** [review]

**Pull request merged**